### ISSUE TITLE
Intent to implement: Built-in see-more for amp-list

### ISSUE THREAD

**Issue by cpapazian** [ampproject/amphtml#13575]
# Motivation
Paging for `<amp-list>` is currently limited and requires integration with other AMP components. Adding a "see more" would simplify common use cases, and adding an option to automatically fetch more more elements would allow publishers to build powerful infinitely scrolling lists.

# Proposed design
Add a "load-more" parameter to `<amp-list>` to indicate that more items are available to be fetched. The overflow element will be used to trigger the fetch. Paging rules will be guessed by the component according to these rules:
- If `pageSize` is specified, paging will use a pageSize+offset model where each additional page will have an `offset` parameter incremented by the value specified by `pageSize`.
- Otherwise a `pageNum` parameter will be added and then incremented. 

Alternatively, a `load-more-bookmark` attribute can specified, in which case, a value will be retrieved from the response (based on simple dot notation) to be passed to the next page. For instance, if the first response is: 
```
{
  "items": [ ... ],
  "bookmark": "foo"
}  
```
then `<amp-list load-more-bookmark="bookmark" src="https://some/url">` would request a second page from https://some/url?bookmark=foo.

Additionally, If the `load-more="auto"` then the fetch will be triggered automatically if the component detects that the bottom of the list is near the bottom of the viewport. 

The component will consider the list to be at its end and not fetch more items if:
- `load-more-bookmark` is specified and resolves to a falsey value
-  response contains a truthy value for `_end`.
- the last response is empty.

# Examples
## Simple Case
```
  <amp-list src="https://some/url" load-more>
    <div overflow><button>See More</button</div>
  </amp-list>
```
The simplest case: Tapping the overflow div will load an additional page, appending `page=n` to the query parameters. The url for the 2nd page will be https://some/url?page=2, the 3rd: https://some/url?page=3, etc. The overflow div will be shown as long as the component determines that there is an additional page. 

In the simple case, the list will be considered to be at its end and the overflow div will not be shown if `_end` is set in the json:
```
{
  "items": [...],
  "_end": true
}
```
or if items is empty: 
```
{
  "items": []
}
```
## pageSize+offset paging
```
  <amp-list src="https://some/url?pageSize=20" load-more>
    <div overflow><button>See More</button</div>
  </amp-list>
```
If `pageSize` is specified in the initial src, additional pages will have an incrementing `offset` value specified in the query params. In the above case the 2nd page would be https://some/url?pageSize=20&offset=20, the 3rd: https://some/url?pageSize=20&offset=40, etc.

## Bookmark paging
```
  <amp-list src="https://some/url" load-more load-more-bookmark="next">
    <div overflow><button>See More</button</div>
  </amp-list>
```
If a `load-more-bookmark` attribute is specified, the response will be parsed and the bookmark value will be retrieved. In the example above, if the first response is: 
```
{
  "items": [],
  "next": "page-2"
}
```
then `?next=page-2` will be appended to the src url as: https://some/url?next=page-2. If `next` is not found in the response:
```
{
  "items": []
}
```
or it resolves to a falsey value: 
```
{
  "items": [],
  "nexts": false
}
```
then the list will be considered to be at its end and the overflow element will not be shown.

## Automatic Paging
```
  <amp-list src="https://some/url" load-more="auto">
    <div overflow><button>See More</button</div>
  </amp-list>
```
If the `load-more` attribute is set to `auto`, then the component will load the next page _automatically_ if the bottom of the amp-list component is near the bottom of the viewport. Initially, "near" will be 1.5x viewport height. 

Note that the resize rules will still be followed in the automatic paging case. That means that the overflow div could still be shown in the case that a resize request is rejected.

/cc @choumx @ericlindley-g @aghassemi

**Comment by cpapazian** [comment]
@choumx @aghassemi we discussed this feature last week at the conference. You suggested that infinite scroll be built into the `<amp-list>`, rather than bolted on via `<amp-position-observer>` ... thoughts on whether we can stage that work? Just adding a "low trust" fetch-page action (and it appears to be safe, based on how the resize rules are implemented in `src/service/resources-impl.js`) would solve for our use case, and an optimized solution could be added later. I'm happy to implement this, and help (or possible implement) the follow up task.

**Comment by cpapazian** [comment]
Additionally, If we decide to include the automatic paging/infinite scroll (as part of the first pass or later), there are a few questions about how and when it would fire ... 

- where do we put the position observer? on the container? right after the container? on the last item in the list?
- do we request the next page immediately after the list is rendered if the last item is seen (or similar criteria is met)?
- are there non-scroll events that should trigger the paging logic?

**Comment by aghassemi** [comment]
@cpapazian Thanks for the great write up and interest in collaborating to design and implement improvements to `amp-list`. 

`amp-list` is an important component in AMP and a clean API and easy-to-use features are critical. It may take you, me and @choumx a few iterations to get to a final proposal to run in the design review meeting for the wider community to comment on, but I hope we can move fast in a pace that works for you to unblock your needs.

Regarding your proposal, overall I really like it. My opinions are below in bullet form:

1- I agree that first-class support for this is needed however I would like to restrict it to just "load-more" cases (which also covers infinite scroll since infinite scroll is simply an automatic trigger of the "load more" button). I believe _numeric_ pagination is best left alone to `amp-bind` and the existing way we have. My reasons are:
   - Numeric pagination is a similar concept to server-side filtering and sort ( which are all really the same thing doing a "replace" with different data) and different than "load-more" which does an "append".
   - Having `amp-list` keep the current page number state outside of `amp-state` is a concern as it would not allow cases such as showing "Page n of m" whereas the current approach via `amp-bind` supports that. 
2- Given (1), let's think about the API: what you have in mind pretty much covers it. Few suggestions regarding naming and enabling infinite scroll. 
   a- `<amp-list load-more=<button|scroll> load-more-bookmark=<KeyInJSONForBookmarkValue>"
   b- `<amp-list load-more=<KeyInJSONForBookmarkValue> load-more-trigger-by-scroll>`
   c- `<amp-list load-more=<KeyInJSONForBookmarkValue> infinite-scroll>`
   d- ???
3- I really like the idea of combining "overflow" and "load more" buttons together,  specially having overflow resize and load next chunk at the same time is a great idea.
4- We may want to have a "no-more-items: true" in JSON response rather than relying on empty `items`. We don't want to get to a case where clicking load-more button won't load anything and would just make it disappear.
5- We would need to append a class to the "overflow" button (or amp-list itself) while the list is loading the next chunk so authors can  show a loading icon in-place or on top of the button.
6- I don't think it should be possible to trigger "load-more" via an Action. This removes any need to worry about resizing, whenever "load-more" button is tapped, list is guaranteed to resize. 
7- I still don't think position-observer should be used to implement infinite scroll. A big issue with using position-observer is that it will always be "too late" as it only works when something comes to viewport. With infinite scrolling we want to load the next chuck way ahead of end of list coming to viewport to minimize chances user would see the bottom of the list. Fortunately this would be fairly simple to implement as all we need to do is monitor position of the overflow button and trigger loading of next chuck when it is "x px" from the bottom of the viewport. We can start by x = half viewport height. May even need to make it configurable at some point.
8- Even with infinite scroll enabled, it is still possible for the user to see the "load-more" button in certain situations (i.e. same situations where overflow would be shown)

**Comment by cpapazian** [comment]
Thanks @aghassemi 

My main motivation for the trigger-via-action mechanism was simplicity. It seemed like a way to break up the work (though my tendency to favor simple primitives and composition is also factor). To your specific points:

6. Because of they way the resize rules were set up, and specifically how FocusHistory works, it appears (from reading code and rudimentary tests) that actions originating from elements outside fo the `<amp-list>` container _will not_ trigger a resize by default. That would make it (relatively) safe to include the trigger-via-action.

7. The position observer _could_ be placed anywhere. Though with regard to fine-grained optimization, your point stands.

That said, if it is simple to implement, as you say, I'm on board building it with the first pass. Should we assume that the list will request the next page on scroll when:
- we are _not_ in an overflow state
- and the last element is less than "x px" from the bottom of the viewport (arbitrary value of x for now).

If that's the case, I agree that it will be simple to implement with a built in scroll observer. If we are in an overflow state, clicking the overflow div will request a new page _and_ force a resize (regardless of whether infinite scroll is enabled).

On the other points ... 
1. Are you drawing a distinction between generalized paging vs. just a see-more feature? That sounds fine to me. I can fix the naming according to your suggestions in #2.  However, I do like the feeling that amp-state and amp-list can work together, with amp-list converting to an amp-state backed list when amp-bind runs (for amp-bind based paging, filtering, sorting, etc). I worry that a see-more feature will break that if there's no way to persist the data outside of amp-list. But I think it's fair to keep that out of scope for now.
4+5. 👍

**Comment by cpapazian** [comment]
@aghassemi I've updated the description based on your suggestions. I removed the trigger-via-action stuff and included a way to specify automatic load-more (infinite scroll) directly. I also took another stab at naming, but happy to defer to you and @choumx on that. Let me know if this looks good to run in the design review ...

**Comment by aghassemi** [comment]
@cpapazian Sorry for the delay, responsed below:

> 6. Because of they way the resize rules were set up, and

Aside from resize, there is also the problem that a low-trust action can now cause network requests.

>7. The position observer could be placed anywhere. 

Position observer only reports when it comes into viewport regardless of where it is placed, so it can never be used to load items earlier than end of list coming to viewport which is too late.

Anyways, seems like we are in agreement with this approach.

> I worry that a see-more feature will break that if there's no way to persist the data outside of amp-list. But I think it's fair to keep that out of scope for now.

Agree , we have had discussion around exposing components themselves as part of AMP state so their internal state can be acted on via `amp-bind` but we don't have that infra yet. If we run into cases where we need that with `show-more` we can either add events that can sync the data with `amp-state` or actually implement the generic feature allowing amp-bind see that in components. Out of scope for this sounds right.

I think this looks pretty close for a review. I suggest cleaning up some of the names in the description (e.g. there is still the `multi-page="bookmark` in there) and add one or two examples that show the usage of the API in a more complete form before the review. Makes it easier for folks unfamiliar with the feature to understands during the review.

**Comment by aghassemi** [comment]
@cpapazian thanks for presenting today. Summary of feedback from Design Review meeting:
1- Forgo the defaults for paging and expect full Url to be in the JSON response.
2- Expose the `how-far-from-viewport` as a configuration and limit to a max (maybe 3 as max? with a default of 0.5?). Name for this: `load-more-auto-distance=<number between 0-3>`? 
3- Let's do more bike shedding on the names. /cc @ericlindley-g 

Any thing else I missed?

**Comment by ericlindley-g** [comment]
Thanks @cpapazian for presenting and @aghassemi for the notes!

FWIW I think the names are solid as is, but want to make sure that @spacedino can take a look for feedback from a UX perspective. We avoid the term "infinite scroll" and are pretty directly descriptive, so I'm optimistic we're in good shape.

The proposal for (2) also SGTM.

One additional bit I wanted to bring up was failure states: this is something I'd like to build better into amp-list in general, but I think the failure state for a "load more" action would be different than a failure state on the initial amp-list load. Do we have a plan for how developers can configure that?

**Comment by cpapazian** [comment]
@aghassemi 
4- preserve json responses. should this be a single json response with the extra pages merged into `items`? or should each page be preserved separately?

**Comment by aghassemi** [comment]
@cpapazian oh yeah, thanks. I think just an array of responses would do. Also out of scope until we actually want to handle DOM recycling.

**Comment by cpapazian** [comment]
@ericlindley-g existing failure state for amp-list is the `<div fallback>` which we obviously don't want to use here, as is. i could see
- do nothing, fail silently
- `<div fallback>` with an extra css class to indicate a different failure state
- some sort of limited retry capability

**Comment by ericlindley-g** [comment]
Some way to indicate a different failure state would be ideal, to avoid leaving users hanging — `<div fallback>` with an extra css class sounds like a lightweight way to do that to me. WDYT?

We're looking into doing a bit of a re-thinking around loading and failure states for amp-list generally, so we could, as an alternative, punt for now (just fail silently and hide the loading state).

**Comment by cpapazian** [comment]
> having two templates and use conditional rendering feature of the amp-mustache to pick but worth exploring to see how it can be made easier.

This would be interesting beyond just end-of-list cases. Any list with heterogeneous content would be simpler to implement. The implementation could be pretty simple: the list item object could set some sort of `type` attribute which matches against an expression on the template.

**Comment by aghassemi** [comment]
@cpapazian we met with our UX (@spacedino) regarding some open UX questions, please see below the summary and please add your feedback/questions. 

## UX Meeting Notes

**A) In "auto" mode, what if user scrolls and reaches the end of the list while load more has been triggered but results have not been rendered (e.g. Fetch is in progress)**.
two options:
1- User sees the `loading` indicator.
2- We don't show the `loading` indicator and instead just show the `overflow`. User will have to `tap` to continue.

We believe **option 2** is better UX and ideally it is rare to get to this state with proper prefetching.

**B) Loading mode**
Based on decision on (A), user can _only_ see the loading state when user "taps" the `overflow` button. We have decided to have a default in AMP for the loading state and allow authors to override.

Default state is: 
- AMP will hide the `overflow`
- AMP will show our default loading indicator. (The three dots for now, to be changed)

Authors can replace the loading via:
`<div load-more-loading> My Loading Spinner! </div>`

Additionally we add `load-more-loading` class to `amp-list` as well. To allow authors to maybe shim the list, etc... (non-MVP)

**C) Failure Mode**
Unlike loading state, AMP will not provide a default `failure` UI. By default, in the failure case, the `overflow` gets shown again.

Author can replace the failure mode via:
`<div load-more-failed>Failed :(  <button>try again</button> </div>`
on `tap` of this `div` will do a refetch similar to `overflow`.

Additionally we add `load-more-failed` class to `amp-list` for extra flexibility. (non-MVP)

@spacedino to also design an AmpByExample samples that covers:
1- custom loading
2- failure
3- end of list being different

**Comment by cpapazian** [comment]
@aghassemi this all sounds good. Agree that there is no need to use a loading indicator for "auto" mode, since if used properly, it should be below the viewport. For the default `load-more-loading` div, is this the same as the `placeholder` div?

**Comment by cpapazian** [comment]
@aghassemi @spacedino I started to work on the loading indicator, and I am not sure it makes sense to show a default (dots) loading indicator, even for the version where an overflow div is used to trigger the next page. The reason is that it is not clear _where_ the div should appear. The option are:
- default `load-more-loading` div works as a regular loading div normally does--overlaid against the entire list component. I feel like this option unnecessarily conceals content that has already been loaded. 
- default `load-more-loading` div appears below existing list content. This option won't conceal existing content, but makes less sense, since we don't actually know how big the list page is going to be. Do we attempt to use extra space already allocated to the list? Do we resize to fit the loading indicator? Or do we not show it when there is no extra space? 

I think it makes sense to show/hide a loading indicator if provided, but the default case, it probably makes more sense to skip it. Let me know your thoughts ...

**Comment by aghassemi** [comment]
@cpapazian thanks for the detailed write-up.

We were thinking `load-more-loading` and `load-more-failed` would be replacements for the `overflow` div and the default loading icon would also only overlay the `overflow` div. So essentially:

**Scenario 1:** Author does not provide `load-more-loading` 
When user taps `overflow` div to load more item, our default loading spinner gets overlayed on `overflow` new items get appended, loading spinner goes away.

 **Scenario 2:** Author provides a custom `load-more-loading` 
When user taps `overflow` div to load more item, `overflow` div gets replaced by `load-more-loading` div, new items get appended, `load-more-loading` gets replaced by `overflow` div again.

**Scenario 3:** Author does not provide a `load-more-failed` (regardless of `load-more-loading`) 
When user taps `overflow` div to load more item, either our or custom loading UI appears, list fails to load, `overflow` is displayed again as it was before. (since we don't provide a default failure UI which is different than loading. This patter is the same across other AMP components where we provide default loading but no default failure)

**Scenario 4:** Author does provide a custom `load-more-failed` (regardless of `load-more-loading`) 
When user taps `overflow` div to load more item, either our or custom loading UI appears, list fails to load, `overflow` is then replaced by `load-more-failed` div.

Now so far everything has been about the bottom of the list and this should cover most cases. In addition to these, we may (not for MVP) add similar states as CSS classes on the `amp-list` component itself to allow for more flexibility (example use-case: Shim the list when user taps show more)

Thoughts?

**Comment by cpapazian** [comment]
Gotcha. That makes a lot more sense. Building the loading indicator into the overflow div in the default state solves the default position issue pretty nicely, since that's always going to be included in a manual load-more case.

When user provides custom `load-more-loading` and/or `load-more-loading` divs, they will be on the hook for positioning appropriately.

**Comment by clodisewp** [comment]
@cpapazian  I need to implement this experimental component into my project recently, but I am new in AMP, and your comments above is hard to me to understand. Could you please show me some webpages with more detailed examples in your spare time. Thanks.

**Comment by cathyxz** [comment]
@clodisewp this is actually done in https://github.com/ampproject/amphtml/issues/14060, and a launch PR was recently merged to master. There are docs for this here https://www.ampproject.org/docs/reference/components/amp-list, under the heading `amp-list-load-more`, which I will formally update once the launch PR reaches production.

**Comment by clodisewp** [comment]
> @clodisewp this is actually done in #14060, and a launch PR was recently merged to master. There are docs for this here https://www.ampproject.org/docs/reference/components/amp-list, under the heading `amp-list-load-more`, which I will formally update once the launch PR reaches production.

@cathyxz Thanks for your reply.  According to your description, it is still an experimental component and is not suitable to implement it now.  Am I right?  Because there is still no formal examples in https://ampbyexample.com/components/amp-list/

**Comment by cathyxz** [comment]
Yup. There are currently examples in https://github.com/ampproject/amphtml/blob/86c2ff61de278f7d0757d1e83e5f7c81e2c5716b/test/manual/amp-list/infinite-scroll-1.amp.html
and on https://amp-infinite-scroll.glitch.me/, I will port them to ABE or more likely the updated amp.dev once the https://github.com/ampproject/amphtml/pull/21355 reaches production.

**Comment by clodisewp** [comment]
> Yup. There are currently examples in https://github.com/ampproject/amphtml/blob/master/test/manual/amp-list/infinite-scroll-1.amp.html
> and on https://amp-infinite-scroll.glitch.me/, I will port them to ABE or more likely the updated amp.dev once the #21355 reaches production.

@cathyxz Great!  I am looking forward to the final example like https://amp-infinite-scroll.glitch.me/ on https://ampbyexample.com/components/amp-list/.

**Comment by clodisewp** [comment]
@cathyxz By the way, I want to know who should I ask questions about 'amp-bind' to modify the attribute 'action' of a form, like this: <form [action]="..."></form>. Thank you.

**Comment by clodisewp** [comment]
@cathyxz Thanks for the example 'https://amp-infinite-scroll.glitch.me/' you provided, I have already achieved effect of infinite scroll.  However, I got a new problem: If the 'amp-list' is wrapped by a 'ul' which has the attribute of '[class]', the page show nothing after it is refreshed.

**Comment by cathyxz** [comment]
Hi @clodisewp !

Generally speaking, usage questions about "how do I do X" are good questions to ask on stackoverflow and then link to in the `using-amp` channel of our slack (amphtml.slack.com). 

For your specific question, I believe that the `action` attribute on `<form>` is not bindable based on our list of bindable attributes [here](https://github.com/ampproject/amphtml/blob/cd0069b6ebce7956e98e2dcdaafb0594ab3fe7c3/extensions/amp-bind/0.1/bind-validator.js#L221-L445). But if you have follow ups unrelated to this issue, feel free to follow up on slack with questions, or to open a separate issue if a bug / feature request is applicable. 

I *think* this particular issue can be closed, since @cpapazian 's original load-more PR has merged long ago, and follow up work is being tracked by https://github.com/ampproject/amphtml/issues/14060. Thank you @cpapazian for your hard work setting all of this up!
